br/pkg/lightning/backend/external/merge_v2.go

```diff
@@ -0,0 +1,176 @@
+	), "merge overlapping files")
+	defer func() {
+		task.End(zap.ErrorLevel, err)
+	}()
+
+	rangeGroupSize := 4 * size.GB
+	failpoint.Inject("mockRangeGroupSize", func(val failpoint.Value) {
+		rangeGroupSize = uint64(val.(int))
+	})
+
+	splitter, err := NewRangeSplitter(
+		ctx,
+		dataFiles,
+		statFiles,
+		store,
+		int64(rangeGroupSize),
+		math.MaxInt64,
+		4*1024*1024*1024,
+		math.MaxInt64,
+		checkHotspot,
+	)
+	if err != nil {
+		return err
+	}
+
+	writer := NewWriterBuilder().
+		SetMemorySizeLimit(DefaultMemSizeLimit).
+		SetBlockSize(blockSize).
+		SetPropKeysDistance(propKeysDist).
+		SetPropSizeDistance(propSizeDist).
+		SetOnCloseFunc(onClose).
+		BuildOneFile(
+			store,
+			newFilePrefix,
+			writerID)
+
+	err = writer.Init(ctx, partSize)
+	if err != nil {
+		return nil
+	}
+
+	bufPool := membuf.NewPool()
+	loaded := &memKVsAndBuffers{}
+	curStart := kv.Key(startKey).Clone()
+	var curEnd kv.Key
+	var maxKey kv.Key
+
+	for {
+		endKeyOfGroup, dataFilesOfGroup, statFilesOfGroup, _, err := splitter.SplitOneRangesGroup()
+		if err != nil {
+			return err
+		}
+		curEnd = kv.Key(endKeyOfGroup).Clone()
+		if len(endKeyOfGroup) == 0 {
+			curEnd = kv.Key(endKey).Clone()
+		}
+		now := time.Now()
+		err = readAllData(
+			ctx,
+			store,
+			dataFilesOfGroup,
+			statFilesOfGroup,
+			curStart,
+			curEnd,
+			bufPool,
+			loaded,
+		)
+		if err != nil {
+			return err
+		}
+		logutil.Logger(ctx).Info("reading external storage in MergeOverlappingFiles",
+			zap.Duration("cost time", time.Since(now)))
+		now = time.Now()
+		sorty.MaxGor = uint64(concurrency)
+		sorty.Sort(len(loaded.keys), func(i, k, r, s int) bool {
+			if bytes.Compare(loaded.keys[i], loaded.keys[k]) < 0 { // strict comparator like < or >
+				if r != s {
+					loaded.keys[r], loaded.keys[s] = loaded.keys[s], loaded.keys[r]
+					loaded.values[r], loaded.values[s] = loaded.values[s], loaded.values[r]
+				}
+				return true
+			}
+			return false
+		})
+		logutil.Logger(ctx).Info("sorting in MergeOverlappingFiles",
+			zap.Duration("cost time", time.Since(now)),
+			zap.Any("key len", len(loaded.keys)))
+		now = time.Now()
+		for i, key := range loaded.keys {
+			err = writer.WriteRow(ctx, key, loaded.values[i])
+			if err != nil {
+				return err
+			}
+		}
+		logutil.Logger(ctx).Info("writing in MergeOverlappingFiles",
+			zap.Duration("cost time", time.Since(now)),
+			zap.Any("key len", len(loaded.keys)))
+		maxKey = kv.Key(loaded.keys[len(loaded.keys)-1]).Clone()
+		curStart = curEnd.Clone()
+		loaded.keys = nil
+		loaded.values = nil
+		loaded.memKVBuffers = nil
+
+		if len(endKeyOfGroup) == 0 {
+			break
+		}
+	}
+
+	var stat MultipleFilesStat
+	stat.Filenames = append(stat.Filenames,
+		[2]string{writer.dataFile, writer.statFile})
+
+	stat.build([]kv.Key{startKey}, []kv.Key{curEnd})
+	if onClose != nil {
+		onClose(&WriterSummary{
+			WriterID:           writer.writerID,
+			Seq:                0,
+			Min:                startKey,
```

Comment:
we can also maintain the minKey like maxKey, in case that startKey does not exist.